### PR TITLE
fix: Deduplicate object candidates

### PR DESCRIPTION
**Pull request by loewenheim**
This documents the invariant that `AllObjectCandidates` is always sorted and unique by `(source, location)` and makes sure that it is always maintained.

#skip-changelog

### CODE REVIEW

File: crates/symbolicator-service/src/types/objects.rs [hunk 1]
```diff
@@ -229,6 +222,20 @@ impl AllObjectCandidates {
...
+    pub fn into_inner(self) -> Vec<ObjectCandidate> {
+        self.0
+    }
+}
+
+impl From<Vec<ObjectCandidate>> for AllObjectCandidates {
+    fn from(mut source: Vec<ObjectCandidate>) -> Self {
+        source
+            .sort_by_cached_key(|candidate| (candidate.source.clone(), candidate.location.clone()));
+        source.dedup_by_key(|candidate| (candidate.source.clone(), candidate.location.clone()));
```
Comment (Swatinem) [inline crates/symbolicator-service/src/types/objects.rs:236]: do we really need to clone here?

File: crates/symbolicator-service/src/services/objects/mod.rs [hunk 2]
```diff
@@ -254,19 +254,31 @@ fn create_candidates(sources: &[SourceConfig], lookups: &[FoundMeta]) -> AllObje
     for meta_lookup in lookups.iter() {
         let source_id = meta_lookup.file_source.source_id();
         source_ids.take(source_id);
-        candidates.push(create_candidate_info(meta_lookup));
+        let location = meta_lookup.file_source.uri();
+        if let Err(idx) =
+            candidates.binary_search_by_key(&(source_id, &location), |c| (&c.source, &c.location))
```
Comment (Swatinem) [inline crates/symbolicator-service/src/services/objects/mod.rs:259]: I’m a bit surprised this code here needs special attention. In which cases does the `lookups` contain duplicates?

File: crates/symbolicator-service/src/types/objects.rs (hunk 1 above)
Comment (loewenheim) [reply crates/symbolicator-service/src/types/objects.rs:236]: As far as I can tell, we do, unfortunately. Just lifetime things.

File: crates/symbolicator-service/src/services/objects/mod.rs (hunk 2 above)
Comment (loewenheim) [reply crates/symbolicator-service/src/services/objects/mod.rs:259]: Probably a good idea to find out why.

File: crates/symbolicator-sources/src/remotefile.rs [hunk 3]
```diff
@@ -129,7 +133,14 @@ impl Iterator for SourceLocationIter<'_> {
...
 
-        self.next.pop().map(SourceLocation::new)
+        self.next.pop().and_then(|next| {
+            if !self.previously_returned.contains(&next) {
+                self.previously_returned.insert(next.clone());
+                Some(SourceLocation::new(next))
+            } else {
+                None
+            }
+        })
```
Comment (Swatinem) [inline crates/symbolicator-sources/src/remotefile.rs:143]: I find it a bit weird to maintain a `HashSet` here. Shouldn’t the downstream user make sure that these things are deduplicated?

Comment (loewenheim) [reply crates/symbolicator-sources/src/remotefile.rs:143]: Yeah, that'd be the other way to do it.

File: crates/symbolicator-service/src/services/download/mod.rs [hunk 4]
```diff
@@ -406,6 +406,12 @@ impl DownloadService {
                 SourceConfig::Filesystem(cfg) => check_source!(cfg => FilesystemRemoteFile),
             }
         }
+        remote_files.sort_by_cached_key(|remote_file| {
+            (remote_file.source_id().clone(), remote_file.uri().clone())
+        });
+        remote_files.dedup_by_key(|remote_file| {
+            (remote_file.source_id().clone(), remote_file.uri().clone())
+        });
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply crates/symbolicator-service/src/services/download/mod.rs]: Not at all. At least in the case of `uri` you're correct, the clone is not necessary—there's something wrong with my setup and I don't see lints in my editor right now, but clippy rejected it for exactly this reason.

This is the error I get when I remove the clone from the `source_id`:
```
error: lifetime may not live long enough
   --> crates/symbolicator-service/src/services/download/mod.rs:409:55
    |
409 |         remote_files.sort_by_cached_key(|remote_file| (remote_file.source_id(), remote_file.uri()));
    |                                          ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                          |          |
    |                                          |          return type of closure is (&'2 SourceId, RemoteFileUri)
    |                                          has type `&'1 RemoteFile`
```
I think the problem is the short lifetime of the closure's `remote_file` parameter.

As for where to do this, `list_files` has 3 non-test callsites: `fetch_file_from_all_sources`, `fetch_line_mapping`, and `find`. We'd probably want the deduplication in all those places; I think it's simpler for `list_files` to already return the deduplicated list.

**Review by Swatinem — APPROVED** [review]
very unfortunate that this needs so many clones :-(

**Pull request merged**